Note that any CI failures are unrelated:
```
FAILED tests/components/config/test_automation.py::test_update_automation_config_with_error[automation_config0] - assert {'message': "...ta['action']"} == {'message': "...a['trigger']"}
  Differing items:
  {'message': "Message malformed: required key not provided @ data['action']"} != {'message': "Message malformed: required key not provided @ data['trigger']"}
  Full diff:
  - {'message': "Message malformed: required key not provided @ data['trigger']"}
  ?                                                                    - ^^^^
  + {'message': "Message malformed: required key not provided @ data['action']"}
  ?                                                                   ++  ^^
Error: Process completed with exit code 1.
```

--- a/homeassistant/components/rainbird/__init__.py
+++ b/homeassistant/components/rainbird/__init__.py
@@ -1,6 +1,7 @@
 """Support for Rain Bird Irrigation system LNK WiFi Module."""
 from __future__ import annotations
 
+import asyncio
 import logging
 
 from pyrainbird.async_client import (
@@ -24,6 +25,8 @@ from homeassistant.helpers import discovery
 from homeassistant.helpers.aiohttp_client import async_get_clientsession
 import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers.typing import ConfigType
+
+from .coordinator import RainbirdUpdateCoordinator
 
 CONF_ZONES = "zones"
 
@@ -93,11 +96,11 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the Rain Bird component."""
 
     hass.data[DATA_RAINBIRD] = []
-    success = False
-    for controller_config in config[DOMAIN]:
-        success = success or await _setup_controller(hass, controller_config, config)
 
-    return success
+    tasks = []
+    for controller_config in config[DOMAIN]:
+        tasks.append(_setup_controller(hass, controller_config, config))
+    return all(await asyncio.gather(*tasks))
 
 
 async def _setup_controller(hass, controller_config, config):
@@ -113,13 +116,22 @@ async def _setup_controller(hass, controller_config, config):
         _LOGGER.error("Unable to setup controller: %s", exc)
         return False
     hass.data[DATA_RAINBIRD].append(controller)
+
+    rain_coordinator = RainbirdUpdateCoordinator(hass, controller.get_rain_sensor_state)
+    delay_coordinator = RainbirdUpdateCoordinator(hass, controller.get_rain_delay)
+
     _LOGGER.debug("Rain Bird Controller %d set to: %s", position, server)
     for platform in PLATFORMS:
         discovery.load_platform(
             hass,
             platform,
             DOMAIN,
-            {RAINBIRD_CONTROLLER: position, **controller_config},
+            {
+                RAINBIRD_CONTROLLER: controller,
+                SENSOR_TYPE_RAINSENSOR: rain_coordinator,
+                SENSOR_TYPE_RAINDELAY: delay_coordinator,
+                **controller_config,
+            },
             config,
         )
     return True

--- a/homeassistant/components/rainbird/binary_sensor.py
+++ b/homeassistant/components/rainbird/binary_sensor.py
@@ -3,8 +3,6 @@ from __future__ import annotations
 
 import logging
 
-from pyrainbird.async_client import AsyncRainbirdController
-
 from homeassistant.components.binary_sensor import (
     BinarySensorEntity,
     BinarySensorEntityDescription,
@@ -12,56 +10,48 @@ from homeassistant.components.binary_sensor import (
 from homeassistant.core import HomeAssistant
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
-
-from . import (
-    BINARY_SENSOR_TYPES,
-    DATA_RAINBIRD,
-    RAINBIRD_CONTROLLER,
-    SENSOR_TYPE_RAINDELAY,
-    SENSOR_TYPE_RAINSENSOR,
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
 )
+
+from . import BINARY_SENSOR_TYPES
 
 _LOGGER = logging.getLogger(__name__)
 
 
-def setup_platform(
+async def async_setup_platform(
     hass: HomeAssistant,
     config: ConfigType,
-    add_entities: AddEntitiesCallback,
+    async_add_entities: AddEntitiesCallback,
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
     """Set up a Rain Bird sensor."""
     if discovery_info is None:
         return
 
-    controller = hass.data[DATA_RAINBIRD][discovery_info[RAINBIRD_CONTROLLER]]
-    add_entities(
+    async_add_entities(
         [
-            RainBirdSensor(controller, description)
+            RainBirdSensor(discovery_info[description.key], description)
             for description in BINARY_SENSOR_TYPES
         ],
         True,
     )
 
 
-class RainBirdSensor(BinarySensorEntity):
+class RainBirdSensor(CoordinatorEntity, BinarySensorEntity):
     """A sensor implementation for Rain Bird device."""
 
     def __init__(
         self,
-        controller: AsyncRainbirdController,
+        coordinator: DataUpdateCoordinator,
         description: BinarySensorEntityDescription,
     ) -> None:
         """Initialize the Rain Bird sensor."""
+        super().__init__(coordinator)
         self.entity_description = description
-        self._controller = controller
 
-    async def async_update(self) -> None:
-        """Get the latest data and updates the states."""
-        _LOGGER.debug("Updating sensor: %s", self.name)
-        state = None
-        if self.entity_description.key == SENSOR_TYPE_RAINSENSOR:
-            state = await self._controller.get_rain_sensor_state()
-        elif self.entity_description.key == SENSOR_TYPE_RAINDELAY:
-            state = await self._controller.get_rain_delay()
-        self._attr_is_on = None if state is None else bool(state)
+    @property
+    def is_on(self) -> bool | None:
+        """Return True if entity is on."""
+        return None if self.coordinator.data is None else bool(self.coordinator.data)

--- a/homeassistant/components/rainbird/coordinator.py
+++ b/homeassistant/components/rainbird/coordinator.py
@@ -1,0 +1,47 @@
+"""Update coordinators for rainbird."""
+
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+import datetime
+import logging
+from typing import TypeVar
+
+import async_timeout
+from pyrainbird.async_client import RainbirdApiException
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+TIMEOUT_SECONDS = 20
+UPDATE_INTERVAL = datetime.timedelta(minutes=1)
+
+_LOGGER = logging.getLogger(__name__)
+
+_T = TypeVar("_T")
+
+
+class RainbirdUpdateCoordinator(DataUpdateCoordinator[_T]):
+    """Coordinator for rainbird API calls."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        update_method: Callable[[], Awaitable[_T]],
+    ) -> None:
+        """Initialize ZoneStateUpdateCoordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Rainbird Zones",
+            update_method=update_method,
+            update_interval=UPDATE_INTERVAL,
+        )
+
+    async def _async_update_data(self) -> _T:
+        """Fetch data from API endpoint."""
+        try:
+            async with async_timeout.timeout(TIMEOUT_SECONDS):
+                return await self.update_method()  # type: ignore[misc]
+        except RainbirdApiException as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err

--- a/homeassistant/components/rainbird/sensor.py
+++ b/homeassistant/components/rainbird/sensor.py
@@ -3,28 +3,24 @@ from __future__ import annotations
 
 import logging
 
-from pyrainbird.async_client import AsyncRainbirdController
-
 from homeassistant.components.sensor import SensorEntity, SensorEntityDescription
 from homeassistant.core import HomeAssistant
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
-from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
-
-from . import (
-    DATA_RAINBIRD,
-    RAINBIRD_CONTROLLER,
-    SENSOR_TYPE_RAINDELAY,
-    SENSOR_TYPE_RAINSENSOR,
-    SENSOR_TYPES,
+from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType, StateType
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
 )
+
+from . import SENSOR_TYPES
 
 _LOGGER = logging.getLogger(__name__)
 
 
-def setup_platform(
+async def async_setup_platform(
     hass: HomeAssistant,
     config: ConfigType,
-    add_entities: AddEntitiesCallback,
+    async_add_entities: AddEntitiesCallback,
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
     """Set up a Rain Bird sensor."""
@@ -32,29 +28,28 @@ def setup_platform(
     if discovery_info is None:
         return
 
-    controller = hass.data[DATA_RAINBIRD][discovery_info[RAINBIRD_CONTROLLER]]
-    add_entities(
-        [RainBirdSensor(controller, description) for description in SENSOR_TYPES],
+    async_add_entities(
+        [
+            RainBirdSensor(discovery_info[description.key], description)
+            for description in SENSOR_TYPES
+        ],
         True,
     )
 
 
-class RainBirdSensor(SensorEntity):
+class RainBirdSensor(CoordinatorEntity, SensorEntity):
     """A sensor implementation for Rain Bird device."""
 
     def __init__(
         self,
-        controller: AsyncRainbirdController,
+        coordinator: DataUpdateCoordinator,
         description: SensorEntityDescription,
     ) -> None:
         """Initialize the Rain Bird sensor."""
+        super().__init__(coordinator)
         self.entity_description = description
-        self._controller = controller
 
-    async def async_update(self) -> None:
-        """Get the latest data and updates the states."""
-        _LOGGER.debug("Updating sensor: %s", self.name)
-        if self.entity_description.key == SENSOR_TYPE_RAINSENSOR:
-            self._attr_native_value = await self._controller.get_rain_sensor_state()
-        elif self.entity_description.key == SENSOR_TYPE_RAINDELAY:
-            self._attr_native_value = await self._controller.get_rain_delay()
+    @property
+    def native_value(self) -> StateType:
+        """Return the value reported by the sensor."""
+        return self.coordinator.data

--- a/homeassistant/components/rainbird/switch.py
+++ b/homeassistant/components/rainbird/switch.py
@@ -3,10 +3,8 @@ from __future__ import annotations
 
 import logging
 
-import async_timeout
 from pyrainbird import AvailableStations
 from pyrainbird.async_client import AsyncRainbirdController, RainbirdApiException
-from pyrainbird.data import States
 import voluptuous as vol
 
 from homeassistant.components.switch import SwitchEntity
@@ -16,17 +14,12 @@ from homeassistant.exceptions import ConfigEntryNotReady, PlatformNotReady
 from homeassistant.helpers import config_validation as cv
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
-from homeassistant.helpers.update_coordinator import (
-    CoordinatorEntity,
-    DataUpdateCoordinator,
-    UpdateFailed,
-)
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
 
-from . import CONF_ZONES, DATA_RAINBIRD, DOMAIN, RAINBIRD_CONTROLLER
+from . import CONF_ZONES, DOMAIN, RAINBIRD_CONTROLLER
+from .coordinator import RainbirdUpdateCoordinator
 
 _LOGGER = logging.getLogger(__name__)
-
-TIMEOUT_SECONDS = 20
 
 ATTR_DURATION = "duration"
 
@@ -58,9 +51,7 @@ async def async_setup_platform(
     if discovery_info is None:
         return
 
-    controller: AsyncRainbirdController = hass.data[DATA_RAINBIRD][
-        discovery_info[RAINBIRD_CONTROLLER]
-    ]
+    controller: AsyncRainbirdController = discovery_info[RAINBIRD_CONTROLLER]
     try:
         available_stations: AvailableStations = (
             await controller.get_available_stations()
@@ -69,7 +60,7 @@ async def async_setup_platform(
         raise PlatformNotReady(f"Failed to get stations: {str(err)}") from err
     if not (available_stations and available_stations.stations):
         return
-    coordinator = RainbirdUpdateCoordinator(hass, controller)
+    coordinator = RainbirdUpdateCoordinator(hass, controller.get_zone_states)
     devices = []
     for zone in range(1, available_stations.stations.count + 1):
         if available_stations.stations.active(zone):
@@ -79,6 +70,7 @@ async def async_setup_platform(
             devices.append(
                 RainBirdSwitch(
                     coordinator,
+                    controller,
                     zone,
                     time,
                     name if name else f"Sprinkler {zone}",
@@ -121,40 +113,20 @@ async def async_setup_platform(
     )
 
 
-class RainbirdUpdateCoordinator(DataUpdateCoordinator[States]):
-    """Coordinator for calendar RPC calls that use an efficient sync."""
-
-    def __init__(
-        self,
-        hass: HomeAssistant,
-        rainbird: AsyncRainbirdController,
-    ) -> None:
-        """Create the CalendarSyncUpdateCoordinator."""
-        super().__init__(
-            hass,
-            _LOGGER,
-            name="Rainbird",
-        )
-        self.rainbird = rainbird
-
-    async def _async_update_data(self) -> States | None:
-        """Fetch data from API endpoint."""
-        try:
-            async with async_timeout.timeout(TIMEOUT_SECONDS):
-                return await self.rainbird.get_zone_states()
-        except RainbirdApiException as err:
-            raise UpdateFailed(f"Error communicating with API: {err}") from err
-
-
 class RainBirdSwitch(CoordinatorEntity, SwitchEntity):
     """Representation of a Rain Bird switch."""
 
     def __init__(
-        self, coordinator: RainbirdUpdateCoordinator, zone: int, time: int, name: str
+        self,
+        coordinator: RainbirdUpdateCoordinator,
+        rainbird: AsyncRainbirdController,
+        zone: int,
+        time: int,
+        name: str,
     ) -> None:
         """Initialize a Rain Bird Switch Device."""
         super().__init__(coordinator)
-        self._rainbird = coordinator.rainbird
+        self._rainbird = rainbird
         self._zone = zone
         self._name = name
         self._state = None

--- a/homeassistant/components/rainbird/switch.py
+++ b/homeassistant/components/rainbird/switch.py
@@ -82,7 +82,6 @@ async def async_setup_platform(
     except ConfigEntryNotReady as err:
         raise PlatformNotReady(f"Failed to load zone state: {str(err)}") from err
 
-    # async_add_entities(devices, True)
     async_add_entities(devices)
 
     async def start_irrigation(service: ServiceCall) -> None:

--- a/requirements_test_all.txt
+++ b/requirements_test_all.txt
@@ -1330,7 +1330,7 @@ pyps4-2ndscreen==1.3.1
 pyqwikswitch==0.93
 
 # homeassistant.components.rainbird
-pyrainbird==0.7.0
+pyrainbird==0.7.1
 
 # homeassistant.components.risco
 pyrisco==0.5.7

--- a/tests/components/rainbird/conftest.py
+++ b/tests/components/rainbird/conftest.py
@@ -3,6 +3,7 @@
 from __future__ import annotations
 
 from collections.abc import Awaitable, Callable, Generator
+from typing import Any
 from unittest.mock import patch
 
 from pyrainbird import encryption
@@ -60,16 +61,23 @@ def platforms() -> list[Platform]:
 
 
 @pytest.fixture
+def yaml_config() -> dict[str, Any]:
+    """Fixture for configuration.yaml."""
+    return CONFIG
+
+
+@pytest.fixture
 async def setup_integration(
     hass: HomeAssistant,
     platforms: list[str],
+    yaml_config: dict[str, Any],
 ) -> Generator[ComponentSetup, None, None]:
     """Fixture for setting up the component."""
 
     with patch(f"homeassistant.components.{DOMAIN}.PLATFORMS", platforms):
 
         async def func() -> bool:
-            result = await async_setup_component(hass, DOMAIN, CONFIG)
+            result = await async_setup_component(hass, DOMAIN, yaml_config)
             await hass.async_block_till_done()
             return result
 

--- a/tests/components/rainbird/test_switch.py
+++ b/tests/components/rainbird/test_switch.py
@@ -14,6 +14,8 @@ from .conftest import (
     ACK_ECHO,
     AVAILABLE_STATIONS_RESPONSE,
     EMPTY_STATIONS_RESPONSE,
+    HOST,
+    PASSWORD,
     URL,
     ZONE_3_ON_RESPONSE,
     ZONE_5_ON_RESPONSE,
@@ -257,3 +259,43 @@ async def test_coordinator_unavailable(
         assert await setup_integration()
 
     assert "Failed to load zone state" in caplog.text
+
+
+@pytest.mark.parametrize(
+    "yaml_config",
+    [
+        {
+            DOMAIN: {
+                "host": HOST,
+                "password": PASSWORD,
+                "trigger_time": 360,
+                "zones": {
+                    1: {
+                        "friendly_name": "Garden Sprinkler",
+                    },
+                    2: {
+                        "friendly_name": "Back Yard",
+                    },
+                },
+            }
+        },
+    ],
+)
+async def test_yaml_config(
+    hass: HomeAssistant,
+    setup_integration: ComponentSetup,
+    responses: list[AiohttpClientMockResponse],
+) -> None:
+    """Test switch platform with fake data that creates 7 zones with one enabled."""
+
+    responses.extend(
+        [mock_response(AVAILABLE_STATIONS_RESPONSE), mock_response(ZONE_5_ON_RESPONSE)]
+    )
+
+    assert await setup_integration()
+
+    assert hass.states.get("switch.garden_sprinkler")
+    assert not hass.states.get("switch.sprinkler_1")
+    assert hass.states.get("switch.back_yard")
+    assert not hass.states.get("switch.sprinkler_2")
+    assert hass.states.get("switch.sprinkler_3")

